Starting docs (#17)
labandalambda/fg-makam: docs/technical-overview.md

# fg.makam

This is the first file we worked on and the best starting point for the project. Here we define all the necessary types and constructors to define Featherweight Go's Abstract Syntax Tree.

> Our representation of Featherweight Go is typical of how you represent a language in functional programming languages. Each node has a type associated with it, and each construct in the language has an associated constructor. For example, an expression is represented by the "expr" datatype. For each expression constructor in the calculus, we define a constructor for the "expr" datatype in Makam. Expressions in Go can be struct literals, method calls, etc. So there will be constructors for struct literals and method calls.

Reviewer: I wouldn't put `etc.` if there's only two more cases
```suggestion
Our representation of Featherweight Go is typical of how you represent a language in functional programming languages. Each node has a type associated with it, and each construct in the language has an associated constructor. For example, an expression is represented by the "expr" datatype. For each expression constructor in the calculus, we define a constructor for the "expr" datatype in Makam. Expressions in Go can be struct literals, method calls, field access or type assertions. So there will be a constructor for each of them.
```

Like this

```
expr: type.

method_call : expr -> string -> list expr -> expr.

> struct_lit : string -> list expr -> expr.

Reviewer: I'd put all the four constructs, even if they're not explicitly explained.

```

Which means, ```expr``` is a type, and both ```method_call``` and ```struct_lit``` are constructors for the type.

The representation we use is based on the FG calculus, whose syntax is presented in figure 9, Griesemer et al (insert ref).

If you took a good look at the previous code paragraph, you might have noticed something unexpected. ```struct_lit``` takes a string and a list of expressions in order to construct a new expression. Why is that? In the FG calculus it says that struct literals are composed of a "structure type name" and a list of expressions. What happened to that "structure type name" and what is the string doing there?

> For our implementation, field names and field methods will be represented as Makam strings. An FG type, whether it is ts (struct type) or ti (interface type), or just arbitrary t, will be represented also as a string. This seems confusing but will eventually become much clearer. Hang in there reader! Just a few more paragraphs before everything becomes clearer.

Reviewer: I'd use the following style throughout the text:
 - `code` for makam code
 - *italic* for the calculus from the paper

```suggestion
For our implementation, field names and field methods will be represented as Makam strings. An FG type, whether it is *ts* (struct type) or *ti* (interface type), or just arbitrary *t*, will be represented also as a string. This seems confusing but will eventually become much clearer. Hang in there reader! Just a few more paragraphs before everything makes sense.
```

> Remember that this is a toy implementation that is interesting for educational and prototyping purposes. So the fact that we represent field names with strings is not a serious problem if it simplifies the job.

Reviewer: I'd say types are the most problematic ones.

```suggestion
Remember that this is a toy implementation that is interesting for educational and prototyping purposes. So the fact that we represent field names or types with strings is not a serious problem if it simplifies the job.
```

> Sadly Makam does not have type synonyms, as far as we know. So we simply annotate in comments that we will be using strings to represent some of the language constructs.

Reviewer: ```suggestion
Sadly Makam does not have type synonyms, [for now](https://github.com/astampoulis/makam/issues/100). So we simply annotate in comments that we will be using strings to represent some of the language constructs.
```

```
% field = string.
% method = string.

> % variable ???

Reviewer: This bit was mainly to state that we don't have a variable construct, I think if we leave it here we should explicitly state what `variable ???` means. I would just delete, since it's better to discuss it later on.

```suggestion
```

% ts, ti, t = string.
```

As we did with expressions, we can define more language constructs by just translating (almost literally) from the FG calculus. 

```
method_sig : type.
method_sig : list string -> string -> method_sig.

method_spec : type.
method_spec : string -> method_sig -> method_spec.

type_lit : type.

struct : list (tuple string string) -> type_lit.
interface : list method_spec -> type_lit.
```

> A ```method_sig``` is constructuted by passing a list of types (strings) and a type to its appropiate constructor. We will not be using the method's argument names, we will only be using their types, so to simplify we do not include the names in our representation.

Reviewer: ```suggestion
A ```method_sig``` is constructed by passing a list of types (`list string`) and a type (`string`) to its appropriate constructor. We will not be using the method's argument names, we will only be using their types, so to simplify we do not include them in our representation.
```
